### PR TITLE
Add submodule support to LazyLoader

### PR DESCRIPTION
**Pull request by jacksontj**
This allows for arbitrarily deep submodules.

This should be merged after #20473 (it contains some of those commits, so that it merges nicely)

### CODE REVIEW

**Comment by jacksontj** [comment]
Not sure why that one test failed on the centos box... that passes locally just fine. @cachedout or @rallytime ?

File: salt/client/mixins.py [hunk 1]
```diff
@@ -293,10 +293,7 @@ def low(self, fun, low):
                     continue
                 completed_funcs.append(mod)
                 for global_key, value in six.iteritems(func_globals):
-                    if six.PY3:
-                        self.functions[fun].__globals__[global_key] = value
-                    else:
-                        self.functions[fun].func_globals[global_key] = value  # pylint: disable=incompatible-py3-code
+                    self.functions[fun].__globals__[global_key] = value
```
Comment (s0undt3ch) [inline salt/client/mixins.py]: This is not Py3 compatible... I was compatible how it was.

PR's don't check Py3 compat yet.

Comment (s0undt3ch) [inline salt/client/mixins.py]: Hmmm

Comment (s0undt3ch) [inline salt/client/mixins.py]: Ignore me

Comment (s0undt3ch) [inline salt/client/mixins.py]:
```python
❯ python                                                                                                                                                                    develop fc9f379 ✗
Python 2.7.8 (default, Oct 27 2014, 04:23:15) 
[GCC 4.9.1 20140903 (prerelease)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo():
...     pass
... 
>>> foo.__globals__
{'__builtins__': <module '__builtin__' (built-in)>, '__name__': '__main__', 'foo': <function foo at 0x7f25934407d0>, '__doc__': None, '__package__': None}
>>> 
```

Comment (s0undt3ch) [inline salt/client/mixins.py]: Same for 2.6:
```python
Python 2.6.9 (unknown, Dec 19 2014, 01:08:28) 
[GCC 4.9.2] on linux3
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo():
...     pass
... 
>>> foo.__globals__
{'__builtins__': <module '__builtin__' (built-in)>, '__name__': '__main__', 'foo': <function foo at 0x7f85791bc5f0>, '__doc__': None, '__package__': None}
>>> 
```
And 3:
```python
Python 3.4.2 (default, Jan 12 2015, 11:38:40) 
[GCC 4.9.2 20141224 (prerelease)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo():
...     pass
... 
>>> foo.__globals__
{'foo': <function foo at 0x7f2e61d3ebf8>, '__builtins__': <module 'builtins' (built-in)>, '__doc__': None, '__name__': '__main__', '__spec__': None, '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__package__': None}
>>> 
```

**Comment by jacksontj** [comment]
Looks like I just had some namespace collisions in my tests ;) easy fix

**Comment by jacksontj** [comment]
Thanks to @steverweber for finding this accidental feature in the old loader ;)

**Comment by jacksontj** [comment]
Interesting... the test failure (https://jenkins.saltstack.com/job/salt-pr-rs-cent7-n/1308/testReport/junit/integration.loader.ext_modules/LoaderOverridesTest/test_overridden_internal/) seems to assume that when you override an internal module we *merge* them? The test code (https://github.com/jacksontj/salt/blob/loader/tests/integration/loader/ext_modules.py#L25) seems to assert that test.ping *and* test.recho exist, which would imply some merging of modules. This seems inconsistent with the rest of the loader, since we don't *merge* varying layers of modules-- mostly because the code isn't guaranteed to be compatible.

Does anyone know why this is expected to work this way? Or should I just fix the test case?

cc @cachedout @s0undt3ch

**Comment by cachedout** [comment]
I would like to get @s0undt3ch 's view here, since I know he did some work with module overrides not too long ago.

**Comment by s0undt3ch** [comment]
There has been some discussion about this before, and I think we should add the intended behavior in the module docstring, or at least, in a comment.
Instead of CC'ing multiple core devs, let me CC @saltstack/salt-core. We should decide on an intended behavior and stick with that.

It's expected that multiple modules which rename themselves into a common virtual name, to have the functions they provide merged into that virtual module namespace. We currently don't support ordering so shadowing functions might result in the wrong function being shadowed.

It's also expected that custom modules, completely override the internal module(not merge). This I found out on a previous issue about this behavior(I though it would still merge).

Let's just get some confirmation on the expected and desired behavior....

I think it would be useful for custom modules to choose the desired behavior. So, if the custom module provides an attribute, for example `__loader_behaviour__ = loader.MERGE_BEHAVIOR` or `__loader_behaviour__ = loader.OVERWRITE_BEHAVIOR` we could select the intended behavior.

**Comment by whiteinge** [comment]
Merging weirds me out.

* How do you tell which version of which function is loaded?
* How does merging private functions work? Say `foo.bar()` calls out to `foo._bar_utils()` at runtime, does Python's scoping transparently handle that or will merging affect which helper is called?

Module-level overriding is simple, single-source-of-truth and easy to debug and reason about.

My $0.02.

**Comment by jacksontj** [comment]
IMO it should be consistent (instead of letting the plugins decide). Merging also doesn't make a lot of sense to me because "foo.bar" and "foo.baz" are then potentially from completely seperate modules/code with inconsistent interfaces (potentially). Think about the case where someone creates a "cmdmod.py" which has one of the functions in the core one, it will overwrite just the one function and (IMO) will make a HUGE mess since most of it seems fine.

**Comment by jacksontj** [comment]
@whiteinge I agree with the ordering being the correct way. 

As for import namespace, the modules are imported based on file name not on virtualname. So we shouldn't have namespace collisions there, BUT if you have a module that calls itself on a specific function it won't necessarily call the same one the user would see. For example

```
# base
def test():
    return test2()

def test2():
    return 'foo'

# some overiden module
def test2():
    return 'bar'
```

In this case if you call foo.test() you will get 'foo'  as the return, but if you run foo.test2() you will get 'bar' which (at least to me) is impossibly confusing to the user.

**Comment by s0undt3ch** [comment]
For clarification, merging already happens in Salt's modules, @techhat will tell you on which modules :smile: 

About scoping, probably needs to be [namespaced](https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/cloudstack.py#L49-L60).

The merging aspect if for a better user experience. Ie a user extends an internal module. The alternative would be to provide the extended part in another namespace.

**Comment by s0undt3ch** [comment]
*is

**Comment by jacksontj** [comment]
The way we do module development is if you need to extend a core module you either contribute to it or create another one which cross calls. Making modules which *inject* their code into core ones seems like a bad idea (although we do it now, still doesn't mean it is a good idea).

**Comment by s0undt3ch** [comment]
We're not injecting....

If functions are called with __salt__[mod.func](), there should be no problem.

**Comment by s0undt3ch** [comment]
Comment updated for clarity.

**Comment by s0undt3ch** [comment]
My attribute idea was for custom modules. Currently, these completely override the internal modules(or at least should)

**Comment by s0undt3ch** [comment]
For clarification, the merging happens at the `__salt__` object level.

So, a module calling a private function of it's module, will use that modules private function, not the private function from a merged module...

**Comment by jacksontj** [comment]
@s0undt3ch AFAICT internal modules are treated the same as external ones-- meaning it could merge assuming you put it in with a different name and a __virtual__ over to the same name as an internal module.

To clarify-- this merging only happens for modules which __virtual__ themselves over into a namespace where something exists.

**Comment by jacksontj** [comment]
@s0undt3ch Correct, and I think that makes it *very* confusing for the user (since you would see potentially inconsistent returns)

**Comment by s0undt3ch** [comment]
Yes, only modules which rename themselves.

**Comment by whiteinge** [comment]
@jacksontj thanks for that explanation. Definitely bad juju.

I'm interested to hear why (and why!) some of our module types have merging behavior. Sounds like something we should avoid for simplicity, consistency, and debug-ability.

**Comment by s0undt3ch** [comment]
I think @techhat made use of that feature for the apache module or state. Ie, several modules contribute to an "apache" module???? At least I think that's what we told me last time we chatted about this.

**Comment by s0undt3ch** [comment]
[salt.modules.apache](https://github.com/saltstack/salt/blob/develop/salt/modules/apache.py) and [salt.modules.deb_apache](https://github.com/saltstack/salt/blob/develop/salt/modules/deb_apache.py)...

**Comment by jacksontj** [comment]
Thats interesting.. since the pkg modules are implemented in a different way (each has their own set of functions which they use).

**Comment by akoumjian** [comment]
I don't have any useful points to add, thought I think this is a really
interesting discussion.  Also, you should all find it hilarious that
'merging' like this, also known as moneky-patching, it the modus operandi
in the Ruby community, which is itself a good indicator that it should be
avoided.

**Comment by s0undt3ch** [comment]
We're not monkey patching! :smiley: 

`__salt__` is a dictionary....

**Comment by jacksontj** [comment]
Yes we aren't monkey patching in the traditional sense, but its *very* similar since the __salt__ dict is a mapping of mod.func -> function, so we are messing with (aka monkey-patching) *that* mapping ;)

**Comment by s0undt3ch** [comment]
I guess I'm the only one who considers this a feature....

**Comment by jacksontj** [comment]
From looking at this apache module, it seems that this could have also been implemented without the merging if the deb_apache module did imports (python imports) from the main apache module-- then it could just replace the whole thing instead of messing with the __salt__ dict.

**Comment by whiteinge** [comment]
@s0undt3ch good memory! Thanks for the example.

That is an interesting trick but I wouldn't like to rely on that behavior, especially not for anything more complicated spanning either large modules or more than two modules. Namespaces are cheap and this just doesn't buy much. Plus none of that behavior is tied into the docs. (Which we could also fix, but that more work.)

@akoumjian lol, that _is_ a good indicator.  ^_^

**Comment by jacksontj** [comment]
@s0undt3ch Its definitely a feature, the question is do we need the complexity that the feature brings ;)

IMO its a lot of magic for seemingly small gains.

**Comment by s0undt3ch** [comment]
Well the apache module is a good example of the feature... Debian adds some stuff not commonly available on other distros. So what's better? Duplicate comon functions?

**Comment by jacksontj** [comment]
Don't have to duplicate, you can either import them (previous comment) or use the depends decorator to optionally expose functions https://github.com/saltstack/salt/blob/develop/salt/utils/decorators/__init__.py#L23

**Comment by s0undt3ch** [comment]
Make Debian users use 2 different apache modules?

**Comment by s0undt3ch** [comment]
I'm loosing the battle... @techhat! Where are you!

**Comment by s0undt3ch** [comment]
:smile:

**Comment by s0undt3ch** [comment]
This was kind of an *easter egg* which I think was not intentional when coded.... @thatch45 might know better about it.
As it is, it's also not reliable because the potential function `overwrite` is alphabetically ordered, or not ordered at all... It just happens to have worked so far...

**Comment by s0undt3ch** [comment]
Let's decide and stick with what we decide... Either improve existing behavior(which extra loader complexity), or drop the merge behavior...

**Comment by s0undt3ch** [comment]
*with

**Comment by jacksontj** [comment]
Yea, so if we want to keep it around I think we have to put some serious time into figuring out exactly how it works (and ensure that it is consistent). As mentioned I'm just not sure its worth the effort. If the intent is to expose functions that are only *sometimes* useful, then we should use a decorator or something to do so (IMO).

**Comment by jacksontj** [comment]
@s0undt3ch Agree, we should pick one and do it-- its current incarnation is a bit of an accidental feature (since its not documented or well understood).

**Comment by s0undt3ch** [comment]
We could also import and namespace the functions... (Like we do for cloud modules)

**Comment by s0undt3ch** [comment]
salt-master(@thatch45) please decide :smile:

**Comment by jacksontj** [comment]
IMO the plethora of less-magic solutions means we should do those and *not* this merging thing ;) 

Assuming we do want to stop merging, I think I'll need to make that a separate PR, since it would require some work on the apache modules (and who knows what else).

**Comment by s0undt3ch** [comment]
Agreed.

**Comment by jacksontj** [comment]
Another way to optionally import functions, you *could* set __load__ (although I think the decorator is nicer still, no bias as the creator of course ;))

**Comment by s0undt3ch** [comment]
I think I need to read that depends decorator docs....

**Comment by jacksontj** [comment]
@s0undt3ch Yea, if we want I can add support for booleans instead of just global namespace checking :)

**Comment by whiteinge** [comment]
I'm ok with having Debian users call two different Apache-related modules since those 'a2*' scripts are so strongly Debian-specific. Avoid duplication by cross-calling other modules via the usual `__salt__` syntax. Since the Debian-specific module requires the main Apache module, make that relationship explicit using `__virtual__()` or the above-mentioned decorator (or similar).

**Comment by jacksontj** [comment]
@whiteinge That works, or we can easily keep them in the same module assuming we use some decorators or something.

**Comment by jacksontj** [comment]
So it sounds like we agree that this *feature* (merging things) is a bad idea? If so, I can open a separate PR with the one commit from here with that.

**Comment by jacksontj** [comment]
Here is a rebase without the "no merge" stuff, #20540

**Comment by jacksontj** [comment]
@cachedout This PR should be safe to merge now, the removal of the merge feature is pulled into a separate PR

**Comment by jacksontj** [comment]
@cachedout And there we go, green! :D

**Comment by cachedout** [comment]
I get out of my afternoon meeting to discover a GH thread that's already gone 57 replies!

You guys rule :]

@jacksontj Merged!

**Comment by thatch45** [comment]
I vote to revert this, the module namespace is short to prevent convoluted applications from overloading the module loader space.....

**Comment by thatch45** [comment]
This will also break the existing nested loader interface that only loads the __init__.py files in subdirs. I am opening a revert PR to discuss this

**Comment by thatch45** [comment]
Let me change my tune, I will let this in if it is config gated and off by default, this is going to break many installs

**Comment by jacksontj** [comment]
@thatch45 This is actually almost completely *identical* to what was there before. Before if I had a dir like:

```
__init__.py
foo.py
bar/baz.py
```
with __init__.py being:
```
import foo
import bar.baz

def ping():
    return foo.SOMETHING(), bar.baz.SOMETHING()
```

In the old mechanism changes to foo would be reloaded, but bar.baz would not (although you could load them and include them).

So, this only *fixes* functionality that already exists. If we want to enforce that directories can only have single files... then why have directories? And if we want to *force* people to not have convoluted directories, why have directories at all?

IMO if we want to keep directory modules we should allow this, otherwise we should remove them.

**Pull request merged**